pkg/assets/containers.go

```diff
@@ -0,0 +1,325 @@
+
+	imageName, err := GetContainerAndRegistryAsString(clusterSpec, imageName)
+
+	if err != nil {
+		glog.Errorf("Unable to get container: %q: %v", imageName, err)
+		return "", fmt.Errorf("unable to parse container %q", imageName)
+	}
+
+	return imageName, nil
+}
+
+const GCR_IO = "gcr.io/google_containers/"
+
+// Cached string for the google container registry.  Usually set to `GCR_IO`.
+var googleRepository *string
+
+// GetGoogleImageRegistryContainer returns a container string, it is used for container that are
+// typically hosted in the google registry.
+func GetGoogleImageRegistryContainer(clusterSpec *kops.ClusterSpec, c string) (string, error) {
+
+	c = strings.TrimPrefix(c, GCR_IO)
+	c, err := GetContainerAsString(c)
+
+	if err != nil {
+		return "", fmt.Errorf("Unable to get google image based container, container does not validate: %v", err)
+	}
+
+	glog.V(2).Infof("container %s", c)
+
+	if googleRepository != nil {
+		glog.V(2).Infof("returning container %s", *googleRepository+c)
+		return *googleRepository + c, nil
+	}
+
+	repo, err := GetRegistryAsString(clusterSpec)
+	glog.V(2).Infof("registry %s", repo)
+
+	if err != nil {
+		return "", fmt.Errorf("Unable to get google image based container: %v", err)
+	}
+
+	if repo != "" {
+		repo = repo + "/"
+		glog.V(2).Infof("repo = %s", repo)
+	} else {
+		glog.V(2).Infof("repo = %s", GCR_IO)
+		repo = GCR_IO
+	}
+
+	googleRepository = &repo
+	glog.V(2).Infof("returning container %s", repo+c)
+
+	return repo + c, nil
+}
+
+// Image returns the docker image name for the specified component
+func Image(component string, clusterSpec *kops.ClusterSpec) (string, error) {
+
+	// TODO figure out if we can add a parameter for version as well
+	// TODO https://github.com/kubernetes/kops/pull/2573#discussion_r117329255
+
+	if component == "kube-dns" {
+		// TODO: Once we are shipping different versions, start to use them
+		i, err := GetGoogleImageRegistryContainer(clusterSpec, "kubedns-amd64:1.3")
+
+		if err != nil {
+			return "", err
+		}
+
+		return i, nil
+
+	}
+
+	if !IsBaseURL(clusterSpec.KubernetesVersion) {
+		c := component + ":" + "v" + clusterSpec.KubernetesVersion
+		i, err := GetGoogleImageRegistryContainer(clusterSpec, c)
+
+		if err != nil {
+			return "", err
+		}
+
+		return i, nil
+
+	}
+
+	baseURL := clusterSpec.KubernetesVersion
+	baseURL = strings.TrimSuffix(baseURL, "/")
+
+	tagURL := baseURL + "/bin/linux/amd64/" + component + ".docker_tag"
+	glog.V(2).Infof("Downloading docker tag for %s from: %s", component, tagURL)
+
+	b, err := vfs.Context.ReadFile(tagURL)
+	if err != nil {
+		return "", fmt.Errorf("error reading tag file %q: %v", tagURL, err)
+	}
+	tag := strings.TrimSpace(string(b))
+	glog.V(2).Infof("Found tag %q for %q", tag, component)
+	c := component + ":" + tag
+
+	i, err := GetGoogleImageRegistryContainer(clusterSpec, c)
+
+	if err != nil {
+		return "", err
+	}
+
+	return i, nil
+}
+
+func IsBaseURL(kubernetesVersion string) bool {
```

Comment:
ahhh ... Every time we do e2e testing :)  There are instructions on how to run kops with a custom k8s bundle.